### PR TITLE
Add speech-to-reverberation modulation energy ratio (SRMR) metric

### PR DESCRIPTION
**Pull request by quancs**
## What does this PR do?

Add SRMR metric

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--1792.org.readthedocs.build/en/1792/

### CODE REVIEW

File: requirements/audio.txt [hunk 1]
```diff
@@ -4,3 +4,5 @@
 # this need to be the same as used inside speechmetrics
 pesq @ git+https://github.com/ludlows/python-pesq
 pystoi <=0.3.3
+Gammatone @ https://github.com/detly/gammatone/archive/master.zip#egg=Gammatone
```
Comment (Borda) [inline requirements/audio.txt:7]: note that this dependency will be skipped for pypi

File: src/torchmetrics/functional/audio/srmr.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+if _TORCHAUDIO_AVAILABEL:
+    from torchaudio.functional.filtering import lfilter
+else:
+    lfilter = None
+    __doctest_skip__ = ["speech_reverberation_modulation_energy_ratio"]
+
+if _GAMMATONE_AVAILABEL:
+    from gammatone.fftweight import fft_gtgram
+    from gammatone.filters import centre_freqs, make_erb_filters
```
Comment (Borda) [inline src/torchmetrics/functional/audio/srmr.py:37]: are these missing in `torchaudio`?

Comment (quancs) [reply src/torchmetrics/functional/audio/srmr.py:37]: Yes, torchaudio doesn't implements these functions.

File: requirements/audio.txt (hunk 1 above)
Comment (quancs) [reply requirements/audio.txt:7]: I tried to use `git+https://github.com/detly/gammatone`, but it failed to be installed in the test environment of `torchmetrics` as it is a conflict dependency with `speechmetrics`.

Comment (quancs) [reply requirements/audio.txt:7]: The `speechmetrics` depends on `https://github.com/detly/gammatone/archive/master.zip#egg=Gammatone`, thus using `git+https://github.com/detly/gammatone` will report a dependency error.

Comment (quancs) [reply requirements/audio.txt:7]: Do you have any ideas to solve it

File: src/torchmetrics/functional/audio/srmr.py (hunk 2 above)
Comment (quancs) [reply src/torchmetrics/functional/audio/srmr.py:37]: Or what do you think about moving the relevant code into torchmetrics

File: src/torchmetrics/audio/srmr.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    As output of `forward` and `compute` the metric returns the following output
+
+    - ``srmr`` (:class:`~torch.Tensor`): float scaler tensor
+
+    .. note:: using this metrics requires you to have ``gammatone`` and ``torchaudio`` installed.
+        Either install as ``pip install torchmetrics[audio]`` or ``pip install torchaudio``
+        and ``pip install git+https://github.com/detly/gammatone``.
+
+    Args:
+        preds: shape ``(..., time)``
```
Comment (SkafteNicki) [inline src/torchmetrics/audio/srmr.py:49]:
```suggestion
```

File: src/torchmetrics/audio/srmr.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        if not _TORCHAUDIO_AVAILABEL or not _GAMMATONE_AVAILABEL:
+            raise ModuleNotFoundError(
+                "speech_reverberation_modulation_energy_ratio requires you to have `gammatone` and"
+                " `torchaudio` installed. Either install as ``pip install torchmetrics[audio]`` or "
+                "``pip install torchaudio`` and ``pip install git+https://github.com/detly/gammatone``"
+            )
+        self.fs = fs
```
Comment (SkafteNicki) [inline src/torchmetrics/audio/srmr.py:101]:
```suggestion
        if not (isinstance(fs, int) and fs > 0):
             raise ValueError(f"Expected argument `fs` to be an int larger than 0, but got {fs}")
        self.fs = fs
```

File: src/torchmetrics/audio/srmr.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        self.n_cochlear_filters = n_cochlear_filters
```
Comment (SkafteNicki) [inline src/torchmetrics/audio/srmr.py:102]:
```suggestion
        if not (isinstance(n_cochlear_filters, int) and n_cochlear_filters > 0):
             raise ValueError(f"Expected argument `n_cochlear_filters` to be an int larger than 0, but got {n_cochlear_filters}")
        self.n_cochlear_filters = n_cochlear_filters
```

File: src/torchmetrics/audio/srmr.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        self.low_freq = low_freq
+        self.min_cf = min_cf
+        self.max_cf = max_cf
+        self.norm = norm
+        self.fast = fast
```
Comment (SkafteNicki) [inline src/torchmetrics/audio/srmr.py:107]: please also add input validation for the rest :)

File: src/torchmetrics/functional/audio/srmr.py (hunk 2 above)
Comment (SkafteNicki) [reply src/torchmetrics/functional/audio/srmr.py:37]: if there are dependency issues due to `gammatone` not being maintained for 9 years, then yes we should include them in tm and just refer to `gammatone`. If not, then lets just keep this dependency.

File: src/torchmetrics/functional/audio/srmr.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+else:
+    fft_gtgram, centre_freqs, make_erb_filters = None, None, None
+    __doctest_skip__ = ["speech_reverberation_modulation_energy_ratio"]
+
+
+@lru_cache(maxsize=100)
+def _calc_erbs(low_freq: float, fs: int, n_filters: int, device: torch.device) -> Tensor:
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/audio/srmr.py:44]: I know it is not required, but could you please add a one line explanation to these private functions.

File: src/torchmetrics/functional/audio/srmr.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+    bw = erbs[k90perc_idx]
+
+    temp = []
+    for b in range(n_batch):
+        score = _cal_srmr_score(bw[b], avg_energy[b], cutoffs=cutoffs)
+        temp.append(score)
+    score = torch.stack(temp)
+
+    if len(shape) > 1:  # recover original shape
+        score = score.reshape(*shape[:-1])
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/audio/srmr.py:295]:
```suggestion
    score = score.reshape(*shape[:-1]) if len(shape) > 1 else score  # recover original shape
        
```

File: tests/unittests/audio/test_srmr.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        #     preds=preds[..., [0], :4000],
+        #     target=preds[..., [0], :4000],
+        #     metric_module=SpeechReverberationModulationEnergyRatioCheat,
+        #     metric_functional=speech_reverberation_modulation_energy_ratio_cheat,
+        #     metric_args={
+        #         "fs": fs,
+        #         "fast": fast,
+        #         "norm": norm
+        #     },
+        # )
```
Comment (SkafteNicki) [inline tests/unittests/audio/test_srmr.py:122]:
```suggestion
```

File: src/torchmetrics/functional/audio/srmr.py (hunk 7 above)
Comment (quancs) [reply src/torchmetrics/functional/audio/srmr.py:44]: Acctually, I don't know the explanation of these functions... And these isn‘t any in the original code.

File: src/torchmetrics/audio/srmr.py (hunk 6 above)
Comment (quancs) [reply src/torchmetrics/audio/srmr.py:119]: added

File: src/torchmetrics/functional/audio/srmr.py (hunk 2 above)
Comment (quancs) [reply src/torchmetrics/functional/audio/srmr.py:37]: I am opening a [pull request](https://github.com/jfsantos/SRMRpy/pull/14) in the upstream trying to fix the [dependency issue](https://github.com/Lightning-AI/torchmetrics/actions/runs/5279262983/jobs/9549684134?pr=1792).

Comment (quancs) [reply src/torchmetrics/functional/audio/srmr.py:37]: The [pull request](https://github.com/jfsantos/SRMRpy/pull/14) has not be merged by the author, so the `gammatone` in requirements is removed temporarily and will be added when it is merged.

File: src/torchmetrics/functional/audio/srmr.py [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+def _cal_srmr_score(bw: Tensor, avg_energy: Tensor, cutoffs: Tensor) -> Tensor:
+    if (cutoffs[4] <= bw) and (cutoffs[5] > bw):
+        kstar = 5
+    elif (cutoffs[5] <= bw) and (cutoffs[6] > bw):
+        kstar = 6
+    elif (cutoffs[6] <= bw) and (cutoffs[7] > bw):
+        kstar = 7
+    elif cutoffs[7] <= bw:
+        kstar = 8
+    return torch.sum(avg_energy[:, :4]) / torch.sum(avg_energy[:, 4:kstar])
```
Comment (vkuznetsovgn) [inline src/torchmetrics/functional/audio/srmr.py:171]: It looks like `kstar` can be referenced before assignment here.

**Review by SkafteNicki — APPROVED** [review]

File: requirements/audio_test.txt [hunk 11]
```diff
@@ -6,3 +6,4 @@ mir-eval >=0.6, <=0.7
 speechmetrics @ git+https://github.com/aliutkus/speechmetrics
 fast-bss-eval >=0.1.0, <0.1.5
 torch_complex <=0.4.3  # needed for fast-bss-eval
+srmrpy @ git+https://github.com/jfsantos/SRMRpy
```
Comment (Borda) [inline requirements/audio_test.txt:9]: note this wont be installed with pypi

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
for the record of https://github.com/detly/gammatone/issues/16#issuecomment-1612917178

**Comment by Borda** [comment]
@SkafteNicki @quancs have you checked the testing of the added package?
ref: https://github.com/jfsantos/SRMRpy/pull/15#issuecomment-1613467903

**Comment by quancs** [comment]
> 



> @SkafteNicki @quancs have you checked the testing of the added package? ref: [jfsantos/SRMRpy#15 (comment)](https://github.com/jfsantos/SRMRpy/pull/15#issuecomment-1613467903)

I tested just now. but the tests didn't get passed.

File: requirements/audio_test.txt (hunk 11 above)
Comment (Borda) [reply requirements/audio_test.txt:9]: Turned out this repo is not tested, so we shall not rely with parity on their numbers
ref: https://github.com/jfsantos/SRMRpy/pull/15

**Comment by Borda** [comment]
@quancs lets add a note that this metric is experimental as we do not have any parity or own reference values to test with

**Comment by quancs** [comment]
> @quancs lets add a note that this metric is experimental as we do not have any parity or own reference values to test with

OK.
